### PR TITLE
Kv processor

### LINKED ISSUE #507: Support parsing messages with key-value strings such as queries and properties
Data Prepper should have a processor which can parse key-value strings from one field and save the results into another field as a map.

For example, on the input event:

```
{
  "query": "key1:value1&key2:value2&key3:value3"
}
```

Could update the event to include a field:

```
{
  "query_params" : {
    "key1" : "value1",
    "key2" : "value2",
    "key3" : "value3"
  }
}
```

This processor must allow configuration of:

* The source field (`query` in the example above)
* The destination field (`query_params` in the example above)
* The delimiter between fields (`&` in the example above)
* The delimiter between key and value (`:` in the example above)
* non_match value - default `null`, the value to assign a key when it has no match
* prefix - A prefix to all keys. default `""`. 

It should also support regex expressions for both types of delimiters.

### PR DESCRIPTION
**Pull request by dapowers87**
### Description
This is the creation of the KeyValueProcessor and its associated tests
 
### Issues Resolved
#507 
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+        assertThat(editedRecords.size(), equalTo(1));
+        assertThat(editedRecords.get(0), notNullValue());
+        assertThat(editedRecords.get(0).getData().toJsonString(),
+                equalTo("{\"message\":\"key1=value1&key2=value2\",\"parsed_message\":{\"key1\":\"value1\",\"key2\":\"value2\"}}"));
+    }
+
+    @Test
+    void testSingleRegexFieldDelimiterKvToObjectKeyValueProcessor() {
+        keyValueProcessor = new KeyValueProcessor(pluginMetrics, getConfig("", null, "!_*!", "=", "", ""));
```
Comment (dlvenable) [inline data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java:65]: This is hard to follow because you pass in quite a few parameters.

File: data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+            field.set(keyValueProcessorConfig, value);
+        } finally {
+            field.setAccessible(false);
+        }
+    }
+
+    private KeyValueProcessorConfig getConfig(final String prefix, final String nonMatchValue,
+                                              final String fieldDelimiterRegex, final String keyValueDelimiterRegex, final String trimKeyRegex,
+                                              final String trimValueRegex) {
+        KeyValueProcessorConfig config = new KeyValueProcessorConfig();
```
Comment (dlvenable) [inline data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java:213]: This would be better with a mock object. Then you don't need to use reflection (Mockito will).

File: data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        } catch (NoSuchFieldException e) {
+            e.printStackTrace();
+        } catch (IllegalAccessException e) {
+            e.printStackTrace();
+        }
+
+        return config;
+    }
+
+    static Record<Event> buildRecordWithEvent(final Map<String, Object> data) {
```
Comment (dlvenable) [inline data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java:231]: You could improve these tests by using a mock here. All of your tests have to assert JSON string values. This can be brittle depending on how it serializes.

If you used Mockito's `verify` functionality, you can ensure that the correct values are _set_ on the Event, rather than checking that the Event itself is complete.

File: data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        assertThat(editedRecords.size(), equalTo(1));
+        assertThat(editedRecords.get(0), notNullValue());
+        assertThat(editedRecords.get(0).getData().toJsonString(),
+            equalTo("{\"message\":\"key1=value1!_____!key2=value2\",\"parsed_message\":{\"key1\":\"value1\",\"key2\":\"value2\"}}"));
+    }
+
+    @Test
+    void testSingleRegexKvDelimiterKvToObjectKeyValueProcessor() {
+        //override setup
+        keyValueProcessor = new KeyValueProcessor(pluginMetrics, getConfig("", null, "&", ":\\+*:", "", ""));
```
Comment (dlvenable) [inline data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java:82]: I recommend using the `createObjectUnderTest()` approach. You create an "object under test" in the `@BeforeEach`, but some tests use it and some don't. This can be error prone as the file changes over time.

I suggest that each test create its own object-under-test and there is no object-under-test field.

File: data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.CoreMatchers.notNullValue;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+public class KeyValueProcessorTests {
```
Comment (dlvenable) [inline data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java:26]: These are good test cases! Thanks! I did make some comments on how you implement them, but I think you've done a good job covering different scenarios.

File: data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+                final String[] terms = group.split(keyValueProcessorConfig.getKeyValueDelimiterRegex(), 2);
+                String key = terms[0];
+                String value;
+
+                if(keyValueProcessorConfig.getTrimKeyRegex() != null && !Objects.equals(keyValueProcessorConfig.getTrimKeyRegex(), "")) {
+                    key = key.replaceAll(keyValueProcessorConfig.getTrimKeyRegex(), "");
+                }
+                key = keyValueProcessorConfig.getPrefix() + key;
+
+                //Expected number of terms to be produced
```
Comment (dlvenable) [inline data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java:54]: I don't know if this comment adds much value.

File: data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorConfig.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    @JsonProperty("field_delimiter_regex")
+    @NotEmpty
+    private String fieldDelimiterRegex = DEFAULT_FIELD_DELIMITER_REGEX;
+
+    @JsonProperty("key_value_delimiter_regex")
+    @NotEmpty
+    private String keyValueDelimiterRegex = DEFAULT_KEY_VALUE_DELIMITER_REGEX;
+
+    @JsonProperty("non_match_value")
+    private String nonMatchValue = DEFAULT_NON_MATCH_VALUE;
```
Comment (dlvenable) [inline data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorConfig.java:37]: I think this should be an `Object`. This way, the value could be `true` instead of `"true"`.

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    private KeyValueProcessor keyValueProcessor;
+    private PluginMetrics pluginMetrics = PluginMetrics.fromNames("kv", "pipeline");
```
Comment (graytaylor0) [inline data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java]: You can mock `PluginMetrics` here. It will also make it easier to test when you add some metrics for this processor

**Review by graytaylor0 — CHANGES_REQUESTED** [review]

File: data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java (hunk 8 above)
Comment (dapowers87) [reply data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java]: You may be looking at an out-of-date version of the tests. I changed it to be mocked before your comment.

File: data-prepper-plugins/key-value-processor/README.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+Create the following file named `logs_json.log` and replace the `path` in the file source of your `pipeline.yaml` with the path of this file.
+
+```json
+{"message": "key1=value&key2=value2"}
+```
+
+When run, the processor will parse the message into the following output:
+
+```json
+{"message": "key1=value&key2=value2", "destination": {"key1": "value1", "key2": "value2"}}
```
[the remark this answers is not in the document]
Comment (dapowers87) [reply data-prepper-plugins/key-value-processor/README.md:28]: Yes, nice catch

File: data-prepper-plugins/key-value-processor/README.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+```
+
+##Configuration
+* `source` - The field in the message that will be parsed. 
+  * Default: `message`
+* `destination` - The field the parsed source will be output to.
+  * Default: `parsed_message`
```
[the remark this answers is not in the document]
Comment (dapowers87) [reply data-prepper-plugins/key-value-processor/README.md:35]: I would imagine so, should I explicitly state that?

File: data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorConfig.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.plugins.processor.keyvalue;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import jakarta.validation.constraints.NotEmpty;
+import jakarta.validation.constraints.NotNull;
+
+public class KeyValueProcessorConfig {
```
[the remark this answers is not in the document]
Comment (graytaylor0) [reply data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorConfig.java:12]: The easiest way to do this is to use `java.regex.Pattern` and try to compile the regex pattern. `compile` throws a `PatternSyntaxException` if the regex is invalid, so you can use a try catch in the `KeyValueProcessorConfig` to validate.

File: data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    @Override
+    public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
+        for(final Record<Event> record : records) {
+            final Map<String, Object> parsedMap = new TreeMap<>();
+            final Event recordEvent = record.getData();
+
+            final String groupsRaw = recordEvent.get(keyValueProcessorConfig.getSource(), String.class);
+            final String[] groups = groupsRaw.split(keyValueProcessorConfig.getFieldDelimiterRegex(), 0);
+            for(final String group : groups) {
+                final String[] terms = group.split(keyValueProcessorConfig.getKeyValueDelimiterRegex(), 2);
```
Comment (graytaylor0) [inline data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java:45]: I think we should look at alternatives to not use `String.split`. This method will be called a lot in this processor, and under the hood it will compile the regex patterns every single time it is called. Ideally, we should compile all regex patterns on construction of the processor.

File: data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java (hunk 8 above)
Comment (graytaylor0) [reply data-prepper-plugins/key-value-processor/src/test/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessorTests.java]: I think you had already made this change before I finished my review. Nice!

File: data-prepper-plugins/key-value-processor/README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (graytaylor0) [reply data-prepper-plugins/key-value-processor/README.md:35]: Yes I think that would be a good thing to clarify here.

File: data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+
+@DataPrepperPlugin(name = "kv", pluginType = Processor.class, pluginConfigurationType = KeyValueProcessorConfig.class)
```
Comment (dlvenable) [inline data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java:24]: I'd prefer to use more robust names. We've been taking that approach so far in Data Prepper. Perhaps `key_value`

File: data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java (hunk 12 above)
Comment (dlvenable) [reply data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java:45]: This should work: https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html#split%28java.lang.CharSequence%29

Yes, we should pre-compile all the regex strings in the constructor.

File: data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+
+    private final Pattern fieldDelimiterPattern;
+    private final Pattern keyValueDelimiterPattern;
+
+    @DataPrepperPluginConstructor
+    public KeyValueProcessor(final PluginMetrics pluginMetrics, final KeyValueProcessorConfig keyValueProcessorConfig) throws PatternSyntaxException {
+        super(pluginMetrics);
+        this.keyValueProcessorConfig = keyValueProcessorConfig;
+
+        if(!validateRegex(keyValueProcessorConfig.getFieldDelimiterRegex())) {
```
Comment (dlvenable) [inline data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java:40]: I think there is an opportunity to improve the errors by validating these in the configuration model. You could create an issue to follow-up on it if we want to approve this PR and fix it later.

File: data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+            final Object existentValue = parsedMap.get(key);
+            if (existentValue instanceof String) {
+                LinkedList<Object> list = new LinkedList<>();
+                list.add(existentValue);
+                list.add(value);
+
+                parsedMap.replace(key, list);
+            } else {
+                ((LinkedList<Object>) existentValue).add(value);
+            }
```
[the remark this answers is not in the document]
Comment (dlvenable) [reply data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java]: Thanks for catching that! I believe the original implementation would have worked for `String`, but some suggestions on default values have changed that.

**Comment by dapowers87** [comment]
#880 has been created to track the improvement of the configuration validation in the KeyValueProcessor

**Review by dlvenable — APPROVED** [review]
This looks good. I'm fine with circling back on the validation and finding other areas for pre-compiling regex expressions.

File: data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (cmanning09) [reply data-prepper-plugins/key-value-processor/src/main/java/com/amazon/dataprepper/plugins/processor/keyvalue/KeyValueProcessor.java]: I think there are some more unit tests that can be added to ensure this functionality is supported.

**Review by graytaylor0 — APPROVED** [review]
